### ISSUE TITLE
Different outputs of the bn layer when converting ResNet from caffe

### ISSUE THREAD

**Issue by nicklhy** [apache/incubator-mxnet#3138]
I am trying to convert the pre-trained ResNet-50 network from [deep-residual-networks](https://github.com/KaimingHe/deep-residual-networks) to mxnet. But it turns out that the predicted results are always **slightly** different from the original one. Then, I found the outputs of a BN layer between caffe and mxnet are always different.
Since caffe's BN layer doesn't really learn gamma/beta, I copied the parameters from the Scale layer to mxnet BN operators:

```
aux_params_map[bn_layer+'_moving_mean'] = mx.nd.array(caffe_bn_blobs[0].data)
aux_params_map[bn_layer+'_moving_var'] = mx.nd.array(caffe_bn_blobs[1].data)
arg_params_map[bn_layer+'_gamma'] = mx.nd.array(caffe_scale_blobs[0].data)
arg_params_map[bn_layer+'_beta'] = mx.nd.array(caffe_scale_blobs[1].data)
```

I set use_global_stats=True and also tried fix_gamma=True/False, but the outputs are still different
Any help would be appreciated.

**Comment by piiswrong** [comment]
Are you using cudnn_bn for both caffe and mxnet? With cudnn_bn the results should be the same. Without cudnn_bn, slight difference due to numeric error is expected.

**Comment by nicklhy** [comment]
@piiswrong , I tested the code both in cpu and gpu mode and the results are similar(produce different outputs). Thus, at least in the cpu mode, I assume the difference may come from the numeric error you mentioned.
However, the difference is actually not that "slight". It often leads to different prediction results. I am not sure if it can really be called a normal numeric error.
For instance, if I feed a image of duck into these two models, the predicted top 5 classes are just slightly different:
```
Caffe's output of the top 5 targets:
rank 1(prob=0.9066): n01847000 drake
rank 2(prob=0.0348): n01855672 goose
rank 3(prob=0.0158): n02018207 American coot, marsh hen, mud hen, water hen, Fulica americana
rank 4(prob=0.0092): n01855032 red-breasted merganser, Mergus serrator
rank 5(prob=0.0055): n01806143 peacock
-------------------------------------------------------------
Mxnet's output of the top 5 targets:
rank 1(prob=0.8810): n01847000 drake
rank 2(prob=0.0369): n01855672 goose
rank 3(prob=0.0153): n02018207 American coot, marsh hen, mud hen, water hen, Fulica americana
rank 4(prob=0.0120): n01806143 peacock
rank 5(prob=0.0083): n01855032 red-breasted merganser, Mergus serrator
```
In another case that I use a chick image, the outputs look more different:
```
Caffe's output of the top 5 targets:
rank 1(prob=0.2092): n01580077 jay
rank 2(prob=0.1190): n02120079 Arctic fox, white fox, Alopex lagopus
rank 3(prob=0.1132): n01537544 indigo bunting, indigo finch, indigo bird, Passerina cyanea
rank 4(prob=0.0893): n02326432 hare
rank 5(prob=0.0614): n01514859 hen
-------------------------------------------------------------
Mxnet's output of the top 5 targets:
rank 1(prob=0.3326): n01580077 jay
rank 2(prob=0.1572): n02120079 Arctic fox, white fox, Alopex lagopus
rank 3(prob=0.0922): n02442845 mink
rank 4(prob=0.0514): n02326432 hare
rank 5(prob=0.0382): n02441942 weasel

```

**Comment by taoari** [comment]
For BatchNorm, try to use `fix_gamma=False, eps=1e-5` for the converted symbol.

**Comment by iacopomasi** [comment]
Hi @nicklhy, All,
do you guys have an updated script to convert Caffe ResNet weights to mxnet ones? I already tried [mxnet converter](https://github.com/dmlc/mxnet/tree/master/tools/caffe_converter) but does not support resnet since mxnet misses some operators.

I get this while converting:

`  File "/home/iac/MXNet/mxnet/tools/caffe_converter/convert_symbol.py", line 169, in proto2script
    raise Exception('Unknown Layer %s!' % layer[i].type)
Exception: Unknown Layer Scale!
`

I have tracked that it seems not a solved issue [here](https://github.com/dmlc/mxnet/issues/676) and [here](https://github.com/dmlc/mxnet/issues/404)

Thanks,
iacopo

**Comment by nicklhy** [comment]
@iacopomasi , the converter in MXNet is not able to convert the resnet [here](https://github.com/KaimingHe/deep-residual-networks) since the different implementations of batch normalization. Kaiming He actually use BN(without learned gamma/beta)+Scale(to learn gamma/beta) in his caffe models, while MXNet can use a single BN to do that.
In my [resnet converter](https://github.com/nicklhy/ResNet_caffe2mxnet), I check the BN+Scale layers manually and replace them with MXNet BN layers. The code can be a little bit dirty due to this and some other inconsistent parameter settings over Kaiming He's resnet 50, 101 and 152. But anyway, you can use it to get the "almost same" MXNet models from the original ones.

**Comment by mli** [comment]
even though i'm agree resnet converter is worth to have, we have pre-trained resnet models on various datasets, including two on 20m images: https://github.com/dmlc/mxnet/tree/master/example/image-classification#pre-trained-models

**Comment by yajiedesign** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
